Capture test output and print it only for failing tests

jarppe opened the pull request:
This PR captures `*out*`, `*err*`, `System/out` and `System/err` for each test thread into a ByteArrayOutputStream. The test reporter requests and prints the output for failing tests.

The implementation has a map `eftest.output-capture/captured-output-contexts` that is keyed by the thread id, and has the output capturing context as a value. The context has a buffer for output and a `PrintWriter` and `PrintStream` instances that both write to the buffer.

Before tests are run, the `*out*`, `*err*`, `System/out` and `System/err` are redirected to a proxied implementations of `PrintWriter` and `PrintStream`. These proxies intercept calls and redirect them to actual writer and stream found from `captured-output-contexts` selected by the current thread.

This allows keeping output from multiple threads separate, even with the `System/out` and `System/err` being shared by all threads.

Since the `PrintStream` and `PrintWriter` are concrete classes without suitable interfaces we can't use `clojure.core/proxy`. This implementation uses the `javassist` library for byte code generation.

The test reporter can then ask the generated output for the current thread and print it.

Deraen commented:
At the least there should probably be an option to enable and disable this. Maybe `:catch-output?` option to `run-tests`?

weavejester commented on eftest/src/eftest/output_capture.clj:
```diff
@@ -0,0 +1,66 @@
+      (.flush writer)
+      (-> buffer .toByteArray (String. StandardCharsets/UTF_8)))))
+
+(defn create-proxy [superclass context-key]
```
Why go to the trouble of creating a proxy? Can't you just have a `PrintStream` and `PrintWriter` write to a string?

Deraen replied:
Proxy is used for `System/setOut` and `setErr`. These are global.

If normal `PrintStream` was used for these, all the output would be written to a single string, and it wouldn't be possible to separate output of different threads. The proxy is used to redirect operations to PrintStream/PrinterWriter based on thread-id.

weavejester replied:
Could you `proxy` an `OuputStream` instead?

```clojure
(proxy [java.io.OutputStream] []
  (write
    ([b]         (.write (get-output-stream) b))
    ([b off len] (.write (get-output-stream) b off len)))))
```

Deraen replied:
`System/setOut` requires specifically `PrintStream`: https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#setOut(java.io.PrintStream)

weavejester replied:
Right, but you could just `(java.io.PrintStream. proxy-output-stream)`.

jarppe replied:
This does not work with multiple threads. Each thread shares the System/out and PrintStream is always buffered, so there is not way to separate the output from different threads.

PrintStream has multiple constructors, but they delegate to these two (private) constructors:

```java
    private PrintStream(boolean autoFlush, OutputStream out) {
        super(out);
        this.autoFlush = autoFlush;
        this.charOut = new OutputStreamWriter(this);
        this.textOut = new BufferedWriter(charOut);
    }

    private PrintStream(boolean autoFlush, OutputStream out, Charset charset) {
        super(out);
        this.autoFlush = autoFlush;
        this.charOut = new OutputStreamWriter(this, charset);
        this.textOut = new BufferedWriter(charOut);
    }
```

So the buffering is pretty tightly built in.

And even without buffering, there is still the problem of separating the outputs from multiple threads.

jarppe replied:
I read this again, you are right, we could implement this "intercepting all method invocations and routing to thread specific streams" without the byte code gen using clojure.lang/proxy, but that would mean that we would have to implement all public methods from PrintStream, all of these:

void println(double)
void println(float)
void println(long)
void println(java.lang.Object)
void println(java.lang.String)
void println([C)
void println()
void println(boolean)
void println(char)
void println(int)
java.lang.Appendable append(java.lang.CharSequence)
java.lang.Appendable append(char)
java.io.PrintStream append(char)
java.lang.Appendable append(java.lang.CharSequence, int, int)
java.io.PrintStream append(java.lang.CharSequence, int, int)
java.io.PrintStream append(java.lang.CharSequence)
java.io.PrintStream format(java.util.Locale, java.lang.String, [Ljava.lang.Object;)
java.io.PrintStream format(java.lang.String, [Ljava.lang.Object;)
void write(int)
void write([B, int, int)
void print(boolean)
void print(long)
void print(double)
void print(float)
void print(java.lang.Object)
void print(java.lang.String)
void print([C)
void print(char)
void print(int)
void flush()
void close()
boolean checkError()
java.io.PrintStream printf(java.lang.String, [Ljava.lang.Object;)
java.io.PrintStream printf(java.util.Locale, java.lang.String, [Ljava.lang.Object;)

weavejester replied:
The buffering of `PrintStream` is a good point. Even with automatic flushing each line, there's still the possibility that two `prints` will hit the same buffer.

As I see it, there are three options:

1. Use the proxy library you were already using
2. Manually enter all 14 distinct methods of PrintStream
3. Use `clojure.reflect/type-reflect` to build a delegating proxy class

I also don't understand why you changed the source to use `proxy` before this discussion was resolved. Is there something that makes proxying the `OutputStream` viable?

weavejester replied:
Ah, nevermind, I just noticed your latest non-review comment.

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -18,15 +19,23 @@
+        test-var      (fn [v]
+                        (capture/begin-capture)
+                        (binding [test/report report
+                                  test/*test-out* *out*
```
Doesn't this override any custom `*test-out*` settings?

jarppe replied:
I'm not sure about that.

weavejester replied:
Could you briefly explain the purpose of this line?

weavejester commented on eftest/src/eftest/runner.clj:
```diff
@@ -18,15 +19,23 @@
            (dorun (->> vars (remove synchronized?) (pmap test))))
-         (doseq [v vars] (each-fixtures #(test-var v))))))))
+         (doseq [v vars] (each-fixtures #(test-var v))))))
+    (capture/restore-capture capture-context)))
```
Can we put this in a `try`-`finally` somehow?

jarppe replied:
Propably. I also considered to put it into the `test-ns` function. The problem is that the context is required in here https://github.com/jarppe/eftest/blob/master/eftest/src/eftest/runner.clj#L28-L29

I'll think about this.

weavejester commented:
Thanks for the PR - I've added some review suggestions and questions.

jarppe commented on eftest/project.clj:
```diff
@@ -1,4 +1,4 @@
-(defproject eftest "0.3.1"
+(defproject eftest "0.3.2-SNAPSHOT"
```
Ok, sorry, I'll update the PR.

jarppe commented on eftest/project.clj:
```diff
@@ -7,4 +7,5 @@
                  [io.aviso/pretty "0.1.33"]
-                 [mvxcvi/puget "1.0.1"]])
+                 [mvxcvi/puget "1.0.1"]
+                 [org.javassist/javassist "3.21.0-GA"]])
```
The problem is the global java.lang.System/out and err. First, they are global and thus shared with all threads. Second, their type is java.io.PrintStream which is a concrete class, not an interface. And, third, it's always buffered.

First and third means that we can't share the PrintStream with multiple threads and still be able to separate what was printed by different threads.

So I came to a conclusion that I need an object that is PrintStream, but would allow capturing the calls to it and reroute the calls to different PrintStreams based on which thread is calling it.

Second point means that standard clojure.lang/proxy and java.lang.reflect.Proxy can't help, byte code generation is required. I know GCLIB, ASM, or Javassist, and I decided to use Javassist, mainly because I'm most comfortable with it.

So the reason for the this dependency is to allow creating an object that "is" PrintStream but allows me to reroute the method invocations.

jarppe commented on eftest/src/eftest/output_capture.clj:
```diff
@@ -0,0 +1,66 @@
+           (java.nio.charset StandardCharsets)
+           (javassist.util.proxy ProxyFactory MethodFilter MethodHandler)))
+
+(set! *warn-on-reflection* true)
```
Ok, I'll clean that too.

jarppe commented:
I reverted the version change and cleaned the set warning for reflection thing.

jarppe commented:
Actually the PrintStream is not buffering the output to next OutputStream (thanks @miikka), so the solution you suggested seems to be working just fine. 

I updated the PR so that the delegating proxy is created on OutputStream using clojure.lang/proxy and I dropped the dep to javassist.

Consequently the code is much shorter and simpler now.

I'm not sure how to test this, but I made a test project with lots of tests that were run in parallel, and I did not see any problems with the outputs.

weavejester commented:
Looking at the source of `PrintStream`, it would appear that all methods write, then flush to the `OutputStream` within a synchronized block, which is why proxying the `OutputStream` is currently safe. While this may change in future (there doesn't seem to be a guarantee in the docs), it seems unlikely to, so it's probably something we can reasonably rely on.

weavejester commented:
Oh, I should also add that I agree with @Deraen that there should be an option to disable this, as it might not be wanted or viable in some cases.

weavejester commented:
This patch should be okay to merge, once an option is added to disable this functionality, and the commits are squashed and the commit message cleaned up.

jarppe commented:
@weavejester Added option for output capture and commits squashed to one.

weavejester commented:
Almost perfect. Could you make it so that `:catch-out?` being `true` is the default? I think, most of the time, catching the output is the right thing to do.

jarppe commented:
The catch-out? is now true by default.

weavejester commented:
Perfect, thanks!
